plugins/font/plugin.js

```diff
@@ -88,131 +193,345 @@
 
-				var previousValue = this.getValue(),
-					style = styles[ value ],
-					previousStyle,
-					range,
-					path,
-					matching,
-					startBoundary,
-					endBoundary,
-					node,
-					bm;
-
-				// When applying one style over another, first remove the previous one (https://dev.ckeditor.com/ticket/12403).
-				// NOTE: This is only a temporary fix. It will be moved to the styles system (https://dev.ckeditor.com/ticket/12687).
-				if ( previousValue && value != previousValue ) {
-					previousStyle = styles[ previousValue ];
-					range = editor.getSelection().getRanges()[ 0 ];
-
-					// If the range is collapsed we can't simply use the editor.removeStyle method
-					// because it will remove the entire element and we want to split it instead.
-					if ( range.collapsed ) {
-						path = editor.elementPath();
-						// Find the style element.
-						matching = path.contains( function( el ) {
-							return previousStyle.checkElementRemovable( el );
-						} );
-
-						if ( matching ) {
-							startBoundary = range.checkBoundaryOfElement( matching, CKEDITOR.START );
-							endBoundary = range.checkBoundaryOfElement( matching, CKEDITOR.END );
-
-							// If we are at both boundaries it means that the element is empty.
-							// Remove it but in a way that we won't lose other empty inline elements inside it.
-							// Example: <p>x<span style="font-size:48px"><em>[]</em></span>x</p>
-							// Result: <p>x<em>[]</em>x</p>
-							if ( startBoundary && endBoundary ) {
-								bm = range.createBookmark();
-								// Replace the element with its children (TODO element.replaceWithChildren).
-								while ( ( node = matching.getFirst() ) ) {
-									node.insertBefore( matching );
-								}
-								matching.remove();
-								range.moveToBookmark( bm );
-
-							// If we are at the boundary of the style element, move out and copy nested styles/elements.
-							} else if ( startBoundary || endBoundary ) {
-								range.moveToPosition( matching, startBoundary ? CKEDITOR.POSITION_BEFORE_START : CKEDITOR.POSITION_AFTER_END );
-								cloneSubtreeIntoRange( range, path.elements.slice(), matching );
-							} else {
-								// Split the element and clone the elements that were in the path
-								// (between the startContainer and the matching element)
-								// into the new place.
-								range.splitElement( matching );
-								range.moveToPosition( matching, CKEDITOR.POSITION_AFTER_END );
-								cloneSubtreeIntoRange( range, path.elements.slice(), matching );
-							}
+			onRender: function() {
+				editor.on( 'selectionChange', function( ev ) {
+					var currentValue = this.getValue(),
+						elementPath = ev.data.path,
+						value = getMatchingValue( editor, elementPath, stylesData );
 
-							editor.getSelection().selectRanges( [ range ] );
+					if ( value ) {
+						if ( value != currentValue ) {
+							this.setValue( value );
 						}
-					} else {
-						editor.removeStyle( previousStyle );
-					}
```

Comment:
The path is refactored. Right now this check together with collapsed selection support is moved to function `hasAlreadyAppliedNewStyle`.

Previously the style was always removed from selection and then applied once again. So subsequent execution of applying the same style result with DOM operation which removed the style and applied the same once again.

To avoid that and not repeating the same command, again and again, it is checked if selection is not styled or contains non-styled fragment. If yes then command is executed, and `removeStyle` operation is the inside of the command.
This generates splits of responsibilities that UI determines if current state of the editor should apply font command or not. And all DOM related operations are made only inside the command.

Reply:
The code which reflecting this else is now located here:
https://github.com/ckeditor/ckeditor4/blob/abb9d6cba0c3dc60706e206f630b479b9f16164a/plugins/font/plugin.js#L114-L116